### ISSUE TITLE
Discussion: Test util for mocking time.sleep()

### ISSUE THREAD

**Issue by bodenr** [jd/tenacity#25]
The way retrying is written and captures time.sleep() in a closure within Retrying.__init__(), consumers cannot patch time.sleep() in their UT code.

To get around this I've been doing some ad-hoc mocking like [1]. While this works I have to replicate it in each openstack project.

Is there some way to can devise for consumers to more easily mockout time.sleep() for their UTs? Maybe there's an easy way to do it as-is and I'm missing it.

[1] https://review.openstack.org/#/c/368235/2/cinder/tests/unit/test_utils.py@38

**Comment by jd** [comment]
Yes, monkey-patching should be the first thing by any project the unit testing. This is correctly done (at least as far as you can correctly monkey patch, sighs) by the `cinder/cmd/*` files, but probably not in the unit tests.

This is a bug that was always there, but just never triggered before. You should probably call `monkey_patch()` in the base test module of Cinder or something like that.

**Comment by bodenr** [comment]
@jd I'm not sure we on the same page here and this issue was not present in retrying...

Forget about ``monkey_patch()`` or anything eventlet related. The problem here is in tenacity's Retrying class's constructor [1] and the fact that it defines ``sleep=time.sleep`` in the actual kwargs. Because its defined in the kwargs, ``time.sleep`` can't be mock patched because it's already captured in a closure. This was not a problem in retrying because retrying didn't define in kwarg.

If I change the tenacity Retrying constructor as follows, all is fine an we can mock patch time.sleep:

```py
    def __init__(self,
                 stop=stop_never, wait=wait_none(),
                 sleep=None,
                 retry=retry_if_exception_type(),
                 before=before_nothing,
                 after=after_nothing):
        # no longer captured in kwargs closure and can be mocked
        self.sleep = sleep or time.sleep
```

Now I realize defining ``time.sleep`` in the kwargs is elegant, but the pain its causing for consumers is unjust IMO.

Thoughts?

[1] https://github.com/jd/tenacity/blob/master/tenacity/__init__.py#L97

**Comment by jd** [comment]
Yeah we're talking about the same thing, but you're wrong when you say "`time.sleep` can't be monkey patch". It can be. It just needs to be monkey patched before `tenacity` is imported, which is currently not the case.

So it is a Cinder problem, not ours.

**Comment by bodenr** [comment]
@jd 
I'm a little confused. I was under the impression that while tenacity intended for general/mass consumption, the primary motivation for creating it was for OpenStack consumption.
So while there are a number of ways to mock patch time with tenacity as-is (mocking time.sleep before importing tenacity, as show in [1] above, etc.), why wouldn't we want to minimize OpenStack porting risk/changes with something as simple as this?

Note that cinder is only 1 example; almost every project using retrying is impacted in this same fashion.

**Comment by jd** [comment]
If creating library primarily for OpenStack consumption means papering over bugs and defect that nobody wants to fix in the projects, this is not my plan. :)

I don't see what's the problem with fixing projects that are miss-using monkey patching. Since this is not a bug in Tenacity, I'm closing.

**Comment by bodenr** [comment]
@jd for the record I'm not understanding why mock patching ``time.sleep()`` (for UTs) is considered a "pappered over bug" here. Perhaps you're aware of something I'm not in which case please share.

As I'm sure you know; what we're going to end up with here is a change in every single project that uses tenacity to globally mock patch ``time.sleep()`` early on, and the provide a way for it's tests to access that mocked sleep object. It doesn't seem like this special handling should be necessary for tenacity, but we obviously disagree here.

**Comment by jd** [comment]
No, actually we agree: I think too that special handling for tenacity should not be needed. :)

What you probably do not understand as you said, is why you have to mock `time.sleep` in e.g. Cinder to have things work and why I call that "papering over the bug".

Cinder relies on eventlet and its monkey patching of `time.sleep()` (and other functions, such as the `threading` module, etc). In order to change the `time.sleep` function by the version from eventlet – which is what that monkey patching thing is about, you have to call the `eventlet.monkey_patch()` function. What it does is basically:

```py
def monkey_patch():
    import time
    time.sleep = _eventlet_sleep
```

You understand that is this `monkey_patch()` is called _after_ `import tenacity`, well, it's too late. If you do:

```py
import tenacity
eventlet.monkey_patch()
```

Then, once tenacity is imported, the `time.sleep` that is tored in the Retrying kwargs is the standard Python `time.sleep` function. And it's exactly the bug you are hitting right now. The code in Cinder should be changed to be:

```py
eventlet.monkey_patch()
import tenacity
```

And that would solve your issue – no need for mock.

You can see that this behaviour is actually implemented in Cinder command line tools (the way Cinder is ran in real life) here:

https://github.com/openstack/cinder/blob/master/cinder/cmd/all.py#L28

Now, clearly the unit tests are not doing that correctly: they don't rely on `cinder.cmd.all` to run and the monkey_patch() function is probably called way after tenacity is loaded, this is why you're hitting this bug, and why you have to `mock` to work around it.

My suggestion here is to fix Cinder (and any other project with that bug) by making sure eventlet.monkey_patch() is called *first thing* before importing anything else and running any other code – like in `cinder.cmd.all`.

I hope it's clearer now! :)

**Comment by bodenr** [comment]
@jd 
Thanks for explaining in greater detail.

However the problem I've been getting at all along is that doing ``eventlet.monkey_patch()`` before importing tenacity doesn't solve the issue.

Cinder already does this early on (before importing tenacity) in it's UTs and ``time.sleep`` cannot subsequently be patched as you'd expect in UTs: https://github.com/openstack/cinder/blob/master/cinder/tests/unit/__init__.py#L30

The way I got this to work is if you change ``cinder.unit.tests.__init__`` to something like:

```py
import time

import eventlet
import mock
from six.moves import builtins

from cinder import objects

eventlet.monkey_patch()
# patch up time sleep before tenacity is imported
mocked_retry_sleep = mock.patch.object(time, 'sleep')
# now consumers can access mocked_retry_sleep if needed
```

So best I can tell, ``time.sleep`` needs to be mock patched before tenacity is imported. This behavior is the reason I opened this issue.

What am I missing here?

**Comment by jd** [comment]
Ok, so I re-read your whole patch and I apologize: I misunderstood your issue. I missed the fact that what Cinder is doing is testing retrying/tenacity itself and checking that `time.sleep` is being called. I had the impression it was because of eventlet (which is a common issue…) but it's actually not related – my bad. I should have dig into your patch a bit more lol.

So, the general issue here is that Cinder is testing retrying/tenacity, which is utterly useless and stupid (yeah I said it lol). I'd suggest to just delete the entire class of test, or move them to tenacity if you/they think the coverage in tenacity is not good enough. I imagine they had test before using retrying and that it was just ported with the test along. Now this is just technical debt.

Now if you really want to test that and see if `time.sleep` have been called using mock (sighs), I'd suggesting just using a custom sleep function as an argument to `tenacity.Retrying` or using the trick you suggested with `mocked_retry_sleep = mock.patch.object(time, 'sleep')`.

But as I said, I'd pick option 1 and just not try to test what is already tested elsewhere. :)

**Comment by bodenr** [comment]
These cinder tests actually test a cinder util function that uses retrying, so I suppose it's debatable if it should really be removed. Regardless; ``time.sleep`` is patched for other reasons and in other projects, not just to test a retried function, so we'll likely need to use the "trick" I mentioned.

I can see you're not going to budge on this one and move ``time.sleep`` out of the constructor; so I'll have to deal with on a per-project basis.

**Comment by harlowja** [comment]
So I to feel that we shouldn't be making `tenacity` paper over some problematic tests in other code-bases (and yes most of those tests shouldn't even exist IMHO, so it'd be great to delete some of them as they are found...); that's somewhat what I would call shoddy engineering (and not exactly a library that I would like to endorse as good quality). Though there is of course the more pragmatic point of view on this, and the fact that its really not that much of a change in the greater picture of things so I think I'd be ok with the change, in the view of it helps in the larger picture being a net-positive.

**Comment by bodenr** [comment]
@harlowja 
Surely I have no "bad engineering" intentions. I think if you spend some time digging into such changes (e.g. [1][2]) you might find it's not so "cut and dry". A number of projects define their own wrapper with customized wait/sleep/etc. -- surely that's game to test to some degree? Other places call a retryable function and have special handling based on the exception type that needs to be tested. Other times retryed functions are just in the callstack of a related tests. In all test instances it's not so easy to say take all those tests out. In fact I don't see what's wrong with patching time.sleep to provide their own additional tests.

I would encourage you or @jd to actually spend some time looking at [1][2].. Perhaps pull down the project and try to port it to using tenacity; as you dig into the details you *might* change your mind on this topic.

[1] https://review.openstack.org/#/c/380552/
[2] https://review.openstack.org/#/c/368235/

**Comment by bodenr** [comment]
Note: its also very possible I'm making things more complex than they need to be. In which case some tips from you guys would be helpful... All the more reason for you to give porting [1] or [2] above to tenacity yourself :)

**Comment by harlowja** [comment]
Right I know u have no "bad engineering" intentions (hope it didn't come off that way). If this is such a small change and it helps some others, meh I'm ok with it (yes I don't think those tests are doing the right thing in the majority of cases, but that's a different story, IMHO it's probably a sign of something else being off design-wise if you need to patch `time.sleep`, but we can debate that one somewhere else I guess, ha).

**Comment by ivanvenosdel** [comment]
I have come across a problem similar to the one @bodenr is describing except with the synchronous version.

@jd @harlowja Yes, ideally projects shouldn't need to test the retries at a level that requires a mock of time.sleep. However it is not uncommon to have complex configuration/custom functions that makes it less straightforward. In that case testing the overall behaviour is necessary if only to prove that our configuration of tenacity will do what we think it will do (e.g. we are not holding it wrong).

Mocking before the underlying code imports tenacity seems like an OK solution but its implementation certainly isn't obvious as having to do something like that is extremely rare IMO. 

Perhaps a good compromise is to add tests to the project that do the kind of mocking of time.sleep you are describing (for both traditional unitest.TestCase and async style tests) and link to it from the documentation.

**Comment by stmcginnis** [comment]
We really do need this. It's not about testing the retry mechanism. It's about being able to test code despite that codes use of retries.

Here's a simple method that we would like to be able to do a basic unit test to validate correct behavior:

https://github.com/openstack/cinder/blob/307fccea4cc9c6496334b0fe137206cb48499bd5/cinder/volume/drivers/inspur/as13000/as13000_driver.py#L776-L786

But since that method is decorated for retries, we are unable to do so without the retries kicking in and causing the test to timeout.

We actually do handle that in that code base. But then found an issue with library code. If a library that is loaded uses tenacity, then the local hack to get around it is too late since tenacity has already been loaded and we have no way of getting test code to skip doing the retries.

**Comment by harlowja** [comment]
I believe u can mock out https://github.com/jd/tenacity/blob/master/tenacity/nap.py#L22 now?

**Comment by stmcginnis** [comment]
Awesome, thanks @harlowja! I wasn't aware of that. It looks like that will work and I can greatly simplify the hack I had in place trying to get around this.

**Comment by harlowja** [comment]
No problemo. Hope all in openstack (opendev?) is doing well!

**Comment by asqui** [comment]
I think this works nicely for cinder because the implementation of `cinder.util.retry()` results in the construction of the `tenacity.Retrying()` object being deferred until the decorated method is called. (Which means that replacing `nap.sleep` any time before the decorated user function is called has the desired effect).

In the more common case where one is using the `tenacity.retry` decorator directly, this will construct a `Retrying` instance immediately, which binds `Retrying.sleep` for that instance to `nap.sleep` at import-time, and precludes any further mocking after that point.

(I discovered this when trying to implement a fix for #228 -- a new issue which I raised recently, not having found this one because it was already closed years ago.)

In any case, I think implementing Option 3 discussed in #228 should allow you to get rid of the complicated work-around here (unless I'm missing something):
https://github.com/openstack/cinder/blob/307fccea4cc9c6496334b0fe137206cb48499bd5/cinder/utils.py#L56
